astropy/cosmology/funcs.py

```diff
@@ -109,37 +139,68 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
     ...            1500 * u.Mpc, zmin=2.5)  # doctest: +FLOAT_CMP
     3.7914913242
 
+    Alternatively the ``bracket`` option can be used to initialise the
+    function solver on a desired region. For the example of angular
+    diameter distance, which has a maximum near a redshift of 1.6 in this
+    cosmology, defining a bracket on either side of this maximum will
+    generally return a solution on the same side.
+
+    >>> z_at_value(Planck13.angular_diameter_distance,
+    ...            1500 * u.Mpc, bracket=(1.0, 1.2))  # doctest: +FLOAT_CMP +IGNORE_WARNINGS
+    0.6812769577
+    >>> z_at_value(Planck13.angular_diameter_distance,
+    ...            1500 * u.Mpc, bracket=(2.0, 2.5))  # doctest: +FLOAT_CMP +IGNORE_WARNINGS
+    3.7914913242
+
     Also note that the luminosity distance and distance modulus (two
     other commonly inverted quantities) are monotonic in flat and open
     universes, but not in closed universes.
     """
-    from scipy.optimize import fminbound
+    from scipy.optimize import minimize_scalar
+
+    opt = {'maxiter': maxfun}
+    # Assume custom methods support the same options as default; otherwise user will see warnings.
+    if isinstance(method, str) and method.lower() == 'bounded':
+        opt['xatol'] = ztol
+        if bracket is not None:
+            warnings.warn(f"Option 'bracket' is ignored by method {method}.")
+    else:
+        opt['xtol'] = ztol
 
     fval_zmin = func(zmin)
     fval_zmax = func(zmax)
     if np.sign(fval - fval_zmin) != np.sign(fval_zmax - fval):
-        warnings.warn("""\
-fval is not bracketed by func(zmin) and func(zmax). This means either
-there is no solution, or that there is more than one solution between
-zmin and zmax satisfying fval = func(z).""")
+        warnings.warn(f"fval is not bracketed by func(zmin)={fval_zmin} and func(zmax)="
+                      f"{fval_zmax}. This means either there is no solution, or that there is "
+                      "more than one solution between zmin and zmax satisfying fval = func(z).")
 
     if isinstance(fval_zmin, Quantity):
         val = fval.to_value(fval_zmin.unit)
-        f = lambda z: abs(func(z).value - val)
     else:
-        f = lambda z: abs(func(z) - fval)
+        val = fval
+
+    # 'Brent' and 'Golden' ignore `bounds`, force solution inside zlim
+    def f(z):
+        if z > zmax:
+            return 1.e300 * (1.0 + z - zmax)
+        elif z < zmin:
+            return 1.e300 * (1.0 + zmin - z)
+        else:
+            return abs(Quantity(func(z)).value - val)
```

Comment:
I think this needs to be fast and Quantity() is not. There probably should be 2 versions of this function, like before.

Reply:
`func` even less – most (all?) of them are integrators over `z`, and the fastest ones I timed, like `Planck18.efunc`, are still 5-6 times slower than casting the result to Quantity. So this might be premature optimisation; otoh it's not completely insignificant, so it may be best to take care of this now before it's forgotten and might become a bottleneck, if the cosmology functions should be further optimised at some point in the future.

Reply:
I like how z out-of-bounds skips the func() evaluation entirely. That seems a very nice optimization.

Reply:
Your idea above to add an extra case for `abs(func(z).value - val)` seems actually the most elegant solution, if that's what it was intended to be.

Reply:
That was my interpretation of the original author's intent. They are very careful throughout the package of stripping units before doing calculations.

Reply:
This appears resolved. 👍